### ISSUE TITLE
Node not completely initialized when consecutive node creation happens 

### ISSUE THREAD

**Issue by gkorland** [FalkorDB/FalkorDB#270]
Created by: @odanoburu
Source: https://github.com/RedisGraph/RedisGraph/issues/2303
Redisgraph 2.8.11, Redis 6.2.6

Minimal example:
```
> GRAPH.QUERY test1 'CREATE (n1:Node) CREATE (n2:Node {n: id(n1)}) RETURN id(n1), n2.n'
1) 1) "id(n1)"
   2) "n2.n"
2) 1) 1) (integer) 2
      2) (integer) -1
3) 1) "Nodes created: 2"
   2) "Properties set: 1"
   3) "Cached execution: 0"
   4) "Query internal execution time: 1.018086 milliseconds"
```

Workaround:
```
> GRAPH.QUERY test1 'CREATE (n1:Node) WITH n1 CREATE (n2:Node {n: id(n1)}) RETURN id(n1), n2.n'
1) 1) "id(n1)"
   2) "n2.n"
2) 1) 1) (integer) 4
      2) (integer) 4
3) 1) "Nodes created: 2"
   2) "Properties set: 1"
   3) "Cached execution: 0"
   4) "Query internal execution time: 0.391228 milliseconds"
```

**Comment by gkorland** [comment]
## Triage: reported symptom is **fixed on C**, but the query now **fails on the Rust engine** for a different reason

**Environment**

| engine | image | digest |
|---|---|---|
| Rust (`main`) | `falkordb/falkordb-server:edge` | `sha256:3624c4b0…` |
| C (reference) | `falkordb/falkordb-server:edge-c` | `sha256:0fc0da0f…` |

### Original symptom — not reproducible on C

```cypher
CREATE (n1:Node) CREATE (n2:Node {n: id(n1)}) RETURN id(n1), n2.n
```

| engine | `id(n1)` | `n2.n` | verdict |
|---|---|---|---|
| RedisGraph 2.8.11 *(as reported)* | 2 | **-1** | ❌ uninitialised node |
| C `edge-c` | 0 | **0** | ✅ **fixed** |
| Rust `main` | — | — | ❌ **`'n1' not defined`** |

The `-1` was `id()` being evaluated against a node whose id had not been assigned yet. On the current C engine `n2.n` correctly matches `id(n1)`, so the reported bug is genuinely fixed there — the `WITH n1` workaround is no longer needed.

### New finding: the Rust engine rejects the query outright

`main` does not return `-1`; it refuses to bind `n1` at all. I narrowed it down — the failure only occurs when a `CREATE`-bound variable is read from an **expression** in a `CREATE` that **directly follows** another `CREATE`:

| query | C | Rust |
|---|---|---|
| `CREATE (a) CREATE (b)-[:R]->(a) RETURN id(b)` | ok | ok ✅ (pattern reference) |
| `CREATE (a) CREATE (b {x: id(a)}) RETURN b.x` | ok | **`'a' not defined`** ❌ |
| `CREATE (a {v:1}) CREATE (b {x: a.v}) RETURN b.x` | ok | **`'a' not defined`** ❌ |
| `CREATE (a) WITH a CREATE (b {x: id(a)}) RETURN b.x` | ok | ok ✅ |
| `CREATE (a) SET a.z = 1 CREATE (b {x: id(a)}) RETURN b.x` | ok | ok ✅ |
| `CREATE (a) MERGE (b {x: id(a)}) RETURN b.x` | ok | ok ✅ |

Any intervening clause resolves it — i.e. the original report's own `WITH n1` workaround is *still* required on Rust, just for a different reason. Filed separately as **#2654**.

Separately, while running these I noticed the `Labels added` statistic diverges between the engines (`CREATE (:Node),(:Node)` reports `2` on C, `1` on Rust; on the 171-node movies dataset C reports `171` and Rust omits the field). Filed as **#2655**.

### Duplicate check

Related scoping issues — none is the same defect:
- #2148 `Node created earlier in a query is not visible inside a later CALL {} subquery` (subquery boundary, not `CREATE`→`CREATE`)
- #1467 `Named path variable in CREATE is not bound` (closed)
- #1279 `CREATE p=(n) RETURN p returns error 'p0' not defined`

**Not a duplicate.**

### Conclusion

- ✅ The reported C-engine defect (`n2.n` = `-1`) is **fixed** — labelling `fixed-rust` is *not* appropriate here, since the Rust engine cannot run the query at all.
- ➡️ Rust follow-on tracked in **#2654**; statistics divergence in **#2655**.
- Recommend **closing this issue as fixed** once #2654 lands, or closing now and tracking Rust behaviour purely in #2654.

*Type set to `Bug`.*

**Comment by gkorland** [comment]
## Re-triaged: original C symptom is fixed — closing as a duplicate of #2654

Verified this issue's exact query against both published engines today:

```cypher
CREATE (n1:Node) CREATE (n2:Node {n: id(n1)}) RETURN id(n1), n2.n
```

| engine | image (digest) | result |
|---|---|---|
| C | `falkordb/falkordb-server:edge-c` (`sha256:0fc0da0f…`) | `id(n1)=0`, `n2.n=0` ✅ — **the reported `n2.n = -1` no longer happens** |
| Rust (`main`) | `falkordb/falkordb-server:edge` (`sha256:3624c4b0…`) | **`'n1' not defined`** ❌ |

So the "node not completely initialized" behaviour described here has been fixed in the C engine. What remains is a *different* failure mode on the Rust engine for the same query: consecutive `CREATE` clauses are fused into a single `Create` op, so the second clause's property expressions are resolved against the scope as it existed before the first `CREATE`, and `n1` is rejected at bind time.

That is tracked with full root-cause analysis, an 8-variant behaviour matrix and the `GRAPH.EXPLAIN` evidence in **#2654**, so I'm marking this as a duplicate of it. The documented workaround here (inserting `WITH n1`) still applies and is confirmed working on both engines.

Please follow #2654 for the fix.

**Comment by gkorland** [comment]
Duplicate of #2654
